File: notifications-api-typescript/src/providers/pushProvider.ts
// Push provider integration

interface ProviderRequest {
  recipient: string;
  message: string;
}

interface ProviderResponse {
  Result: string;
  ErrorCode: string;
  Message: string;
}

export function send(req: ProviderRequest): ProviderResponse {
  if (!req.recipient || !req.message) {
    return {
      Result: "InvalidRequest",
      ErrorCode: "PUSH_INVALID_DEVICE",
      Message: "[push] invalid device token",
    };
  }
  const r = Math.random();
  if (r < 0.45) {
    return {
      Result: "Success",
      ErrorCode: "PUSH_OK",
      Message: "[push] notification delivered",
    };
  }
  if (r < 0.7) {
    const code = Math.random() < 0.5 ? "PUSH_TEMP_001" : "PUSH_TEMP_002";
    return {
      Result: "TemporaryFailure",
      ErrorCode: code,
      Message: "[push] temporary outage, retry later",
    };
  }
  const code = Math.random() < 0.5 ? "PUSH_PERM_REJECTED" : "PUSH_PERM_EXPIRED";
  return {
    Result: "PermanentFailure",
    ErrorCode: code,
    Message: "[push] device token rejected",
  };
}

File: notifications-api-dotnet/Providers/PushProvider.cs
// Push provider integration
namespace NotificationApi.Providers;

public static class PushProvider
{
    public static ProviderResponse Send(Dictionary<string, string> req)
    {
        var recipient = req.GetValueOrDefault("recipient", "");
        var message = req.GetValueOrDefault("message", "");
        if (string.IsNullOrEmpty(recipient) || string.IsNullOrEmpty(message))
        {
            return new ProviderResponse
            {
                Result = "InvalidRequest",
                ErrorCode = "PUSH_INVALID_DEVICE",
                Message = "[push] invalid device token"
            };
        }
        var r = Random.Shared.NextDouble();
        if (r < 0.45)
        {
            return new ProviderResponse
            {
                Result = "Success",
                ErrorCode = "PUSH_OK",
                Message = "[push] notification delivered"
            };
        }
        if (r < 0.7)
        {
            var code = Random.Shared.NextDouble() < 0.5 ? "PUSH_TEMP_001" : "PUSH_TEMP_002";
            return new ProviderResponse
            {
                Result = "TemporaryFailure",
                ErrorCode = code,
                Message = "[push] temporary outage, retry later"
            };
        }
        var permCode = Random.Shared.NextDouble() < 0.5 ? "PUSH_PERM_REJECTED" : "PUSH_PERM_EXPIRED";
        return new ProviderResponse
        {
            Result = "PermanentFailure",
            ErrorCode = permCode,
            Message = "[push] device token rejected"
        };
    }
}
